File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/campinas/CampinasPetcampCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.campinas;

import br.com.lett.crawlernode.core.models.RankingProduct;
import br.com.lett.crawlernode.core.models.RankingProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords;
import br.com.lett.crawlernode.exceptions.MalformedProductException;
import br.com.lett.crawlernode.util.CrawlerUtils;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.UnsupportedEncodingException;
import java.util.Arrays;

public class CampinasPetcampCrawler extends CrawlerRankingKeywords {

   public CampinasPetcampCrawler(Session session) {
      super(session);
   }

   @Override
   protected void extractProductsFromCurrentPage() throws UnsupportedEncodingException, MalformedProductException {
      this.pageSize = 12;

      this.log("Página " + this.currentPage);

      String url = "https://www.petcamp.com.br/" + this.keywordWithoutAccents.replace(" ", "%20") + "#" + this.currentPage;

      this.log("Link onde são feitos os crawlers: " + url);

      this.currentDoc = fetchDocument(url);

      Elements products = this.currentDoc.select(".sawi-shelf.n12colunas > ul > li:nth-child(odd)");

      if (!products.isEmpty()) {
         if (this.totalProducts == 0) {
            setTotalProducts();
         }
         for (Element e : products) {
            String internalId = CrawlerUtils.scrapStringSimpleInfoByAttribute(e, ".sawi-shelf-box > .buy-button-normal", "id");
            String productUrl = CrawlerUtils.scrapUrl(e, ".sawi-shelf-inner > h3 > a", "href", "https:", "www.petcamp.com.br");
            String name = CrawlerUtils.scrapStringSimpleInfoByAttribute(e, ".sawi-shelf-inner > h3 > a", "title");
            String imageUrl = CrawlerUtils.scrapSimplePrimaryImage(e, ".sawi-shelf-image > a > img", Arrays.asList("src"), "https", "www.petcamp.com.br");
            Integer price = CrawlerUtils.scrapPriceInCentsFromHtml(e, ".sawi-shelf-inner > p > .new-price", null, true, ',', session, null);
            boolean isAvailable = price != null;

            RankingProduct productRanking = RankingProductBuilder.create()
               .setUrl(productUrl)
               .setInternalId(internalId)
               .setName(name)
               .setPriceInCents(price)
               .setAvailability(isAvailable)
               .setImageUrl(imageUrl)
               .build();

            saveDataProduct(productRanking);

            if (this.arrayProducts.size() == productsLimit) {
               break;
            }

         }
      } else {
         this.result = false;
         this.log("Keyword sem resultado!");
      }

      this.log("Finalizando Crawler de produtos da página " + this.currentPage + " - até agora " + this.arrayProducts.size() + " produtos crawleados");
   }
   @Override
   protected void setTotalProducts() {
      this.totalProducts = CrawlerUtils.scrapSimpleInteger(currentDoc, ".searchResultsTime:first-child > .resultado-busca-numero > .value", false);
      this.log("Total da busca: " + this.totalProducts);
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/campinas/CampinasPetcampCrawler.kt
package br.com.lett.crawlernode.crawlers.ranking.keywords.campinas

import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords
import br.com.lett.crawlernode.util.CrawlerUtils

class CampinasPetcampCrawler(session: Session?) : CrawlerRankingKeywords(session) {

    override fun extractProductsFromCurrentPage() {
        pageSize = 12
        log("Página $currentPage")

        val url = "https://www.petcamp.com.br/buscapagina?ft=$keywordEncoded&PS=12&sl=5e1350fe-8755-42e7-acd1-7682c25399f5&cc=12&sm=0&PageNumber=$currentPage"
        log("Link onde são feitos os crawlers: $url")

        currentDoc = fetchDocument(url)

        val products = currentDoc.select(".sawi-shelf ul li .sawi-shelf-box")

        if (!products.isEmpty()) {
            if (totalProducts == 0) {
                setTotalProducts()
            }
            for (product in products) {
                val internalId = CrawlerUtils.scrapStringSimpleInfoByAttribute(product, ".buy-button-normal", "id")
                val internalPid = internalId
                val productUrl = CrawlerUtils.scrapStringSimpleInfoByAttribute(product, ".sawi-shelf-image a", "href")
                saveDataProduct(null, null, productUrl)

                log("Position: $position - InternalId: $internalId - InternalPid: $internalPid - Url: $productUrl")
                if (arrayProducts.size == productsLimit) {
                    break
                }
            }
        } else {
            result = false
            log("Keyword sem resultado!")
        }
        log("Finalizando Crawler de produtos da página $currentPage até agora ${arrayProducts.size} produtos crawleados")
    }

   override fun setTotalProducts() {
      val url = "https://www.petcamp.com.br/$keywordEncoded"
      val page = fetchDocument(url)
      val json = CrawlerUtils.selectJsonFromHtml(page, "script", "vtex.events.addData(", ");", false, false)
      totalProducts = json.optInt("siteSearchResults")
      println(totalProducts)
      log("Total da busca: $totalProducts")
   }

}
